src/DssEmergencySpell.sol

```diff
@@ -0,0 +1,135 @@
+}
+
+interface DssAction {
+    function actions() external;
+    function description() external view returns (string memory);
+    function execute() external;
+    function nextCastTime(uint256 eta) external view returns (uint256);
+    function officeHours() external view returns (bool);
+}
+
+interface DssEmergencySpellLike is DssExec, DssAction {
+    function description() external view override(DssExec, DssAction) returns (string memory);
+    function officeHours() external view override(DssExec, DssAction) returns (bool);
+}
+
+abstract contract DssEmergencySpell is DssEmergencySpellLike {
+    /// @dev The chainlog contract reference.
+    ChainlogLike internal constant _log = ChainlogLike(0xdA0Ab1e0017DEbCd72Be8599041a2aa3bA7e740F);
+
+    // @dev The reference to the `pause` contract.
+    address public immutable pause = _log.getAddress("MCD_PAUSE");
+    /// @dev The chainlog address.
+    address public constant log = address(_log);
+    /// @dev In regular spells, `eta` is used to enforce the GSM delay.
+    ///      For emergency spells, the GSM delay is not applicable.
+    uint256 public constant eta = 0;
+    /// @dev Keeping the same value as regular spells.
+    bytes public constant sig = abi.encodeWithSelector(DssAction.execute.selector);
+    /// @dev Emergency spells should not expire.
+    uint256 public constant expiration = type(uint256).max;
+    // @dev Office Hours is always `false` for emergency spells.
+    bool public constant officeHours = false;
+    // @dev `action` is expected to return a valid address.
+    //      We also implement the `DssAction` interface in this contract.
+    address public immutable action = address(this);
+    // @dev An emergency spell can be cast as soon as it is deployed.
+    //      Notice that cast is always a no-op.
+    uint256 internal immutable _nextCastTime = block.timestamp;
+
+    /**
+     * @dev In regular spells, `tag` is an immutable variable with the code hash of the spell action.
+     *      It specifically uses a separate contract for spell action because `tag` is immutable and the code hash of
+     *      the contract being initialized is not accessible in the constructor.
+     *      Since we do not have a separate contract for actions in Emergency Spells, `tag` has to be turned into a
+     *      getter function instead of an immutable variable.
+     * @return The contract codehash.
+     */
+    function tag() external view returns (bytes32) {
+        return address(this).codehash;
+    }
+
+    /**
+     * @notice Triggers the emergency actions of the spell.
+     * @dev Emergency spells are triggered when scheduled.
+     *      This function maintains the name for compatibility with regular spells, however nothing is actually being
+     *      scheduled. Emergency spells take affect immediately, so there is no need to call `pause.plot()`.
+     */
+    function schedule() external {
+        _emergencyActions();
+    }
```

Comment:
Looking at the [chief-keeper block-scheme](https://github.com/makerdao/chief-keeper), I understand that it will try to call `spell.schedule()` in the loop, until "it is scheduled". On the technical level, "scheduled" is [checked via `spell.eta != 0`](https://github.com/makerdao/chief-keeper/blob/fd217f01506e51070f4a706904ea87a4d63465d2/chief_keeper/chief_keeper.py#L328). In case we don't modify `eta`, the keeper will call emergency spell in the loop in every single block and the whole interface "compliancy" wouldn't make much sense.

To solve this loop, we can either:
- A. Always use factories and require deployment, even for multi-variations. Have fully compliment `eta` logic, expiration and one-time usage logic. I hardly can see when emergency spell with the same hat will need to be executed more than once
- B. Set `eta` inside `spell.schedule` to non zero
    - Additional complexity: Setting `eta` will prevent spell from being directly reusable. We can potentially circumvent this by modifying it back to 0 either via `cast` or manually/permissionlessly via some other, e.g. `reset()` method
- C. Modify keeper logic to determine that spell was already scheduled
    - Additional complexity: there no cross-compatible event or state variable that can be used for checking if "a spell" was scheduled. For example, if `plot` would've been called via emergency spell, keeper can check [`pause.plans[hash] != true`](https://github.com/dapphub/ds-pause/blob/0763eafcf926fd2e073aee5f047f3decb842231c/src/pause.sol#L79-L84) instead of `spell.eta != 0`

All in all, I would recommend to make some changes to the base contract to prevent keeper loop.

Reply:
Looking at that line we have:
`if spell.done() == False`

Notice that all emergency spells have meaningful `done()` implementations, meaning that right after `schedule()` is called, it should return `true`, which would prevent the loop in the keeper.

Reply:
> A. Always use factories and require deployment, even for multi-variations. Have fully compliment eta logic, expiration and one-time usage logic. I hardly can see when emergency spell with the same hat will need to be executed more than once

The idea of having pre-deployed spells is to not require any non-engineer to ever be required to provide inputs manually. Even in the cases where there are factories, they would be used by engineers to deploy the contracts and hand over only the addresses to other stakeholders.